---
title: "PowerExplorer Manual"
author: "Xu Qiao, Laura Elo"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
toc: true
vignette: >
  %\VignetteIndexEntry{PowerExplorer Manual}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

\newpage

# Abstract
This vignette demonstrates R package `PowerExplorer` as a power and sample size
estimation tool for RNA-Seq and quantitative proteomics data.

`PowerExplorer` contains the following main features:

- Estimation of power based on the current data
- Prediction of power corresponding to the increased sample sizes
- Result visualizations

# Introduction
Power and sample size estimation is one of the important principles in designing
next-generation sequencing experiments to discover differential expressions.
`PowerExplorer` is a power estimation and prediction tool currently applicable
to RNA-Seq and quantitative proteomics experiments.

The calculation procedure starts with estimating the distribution parameters of
each gene or protein (following referred as feature for simplicity). With the
obtained prior distribution of each feature, a specified amount of simulations
are executed to generate data (read counts for RNA-Seq and protein abundance for
proteomics) repetitively for each entry based on null and alternative hypotheses
. Furthermore, the corresponding statistical tests (t-test or Wald-test) are
performed and the test statistics are collected. Eventually the statistics will
be summarized to calculate the statistical power.

\newpage

# Prepare Input Data

For both RNA-Seq (gene expression levels) and quantitative proteomics (protein
abundance levels) datasets, the data matrix should be arranged as features in
rows and samples in columns, for example:
```{r showDataExample, message=FALSE, warning=FALSE}
library(PowerExplorer)
data("exampleRNASeqData")
head(exampleRNASeqData$dataMatrix[,1:6])
```

A grouping vector indicating the sample groups to which all the samples belong
should also be created, for example:
```{r}
show(exampleProteomicsData$groupVec)
```
The sample groups corresponding to the data:
```{r}
colnames(exampleProteomicsData$dataMatrix)
```

Note that the grouping vector length should be equal to the column number of the
data matrix.

\newpage

# Run Estimation

Here we use a randomly generated RNASeq dataset `exampleRNASeqData` as an
example to estimate the current power of the dataset. The input dataset is named
as `dataMatrix` and the grouping vector as `groupVec`.

To run the estimation, apart from the input, we still need to specify the
following parameters:

- `isLogTransformed`: FALSE; the input data is not log-transformed.
- `dataType`: "RNA-Seq"; the datatype can be declared as "Proteomics" or
"RNA-Seq".
- `minLFC`: 0.5; the threshold of Log2 Fold Change, proteins with lower LFC will
be discarded.
- `alpha`: 0.05; the controlled false positive (Type I Error) rate.
- `ST`: 50; the simulation of each gene will be run 50 times (ST>50 is
  recommended).
- `seed`: 345; optional, a seed value for the random number generator to
maintain the reproducibility.
- `showProcess`: FALSE; no detailed processes will be shown, set to TRUE if
debug is needed.
- `saveSimulatedData`: FALSE; if TRUE, save the simulated data in ./savedData
directory.

The results will be summaried in barplot, boxplot and summary table.

```{r estimation run, fig.keep='none', message=FALSE, warning=FALSE}
library(PowerExplorer)
data("exampleRNASeqData")
res <- estimatePower(inputObject = exampleRNASeqData$dataMatrix,
                               groupVec = exampleRNASeqData$groupVec,
                               isLogTransformed = FALSE,
                               dataType = "RNASeq",
                               minLFC = 0.5,
                               alpha = 0.05,
                               ST = 50,
                               seed = 345)
```


\newpage

## Visualization
The estimated results can be summarized using `plotEstPwr`, the only input
needed is the `estimatedPower`, which should be the estimated power object
returned from  `estimatePower`.

```{r plot estimated power}
plotEstPwr(res)
```

The graph contains 3 plots, the `barplot` vertically shows the number of
genes/proteins above the minLFC threshold, columns indicates the comparison
pairs, each column presents the proportions of three power levels in three
colours as indicated in the legend `power.level`; The boxplot shows the overall
power distribution of each comparsion; And the summary table summarize the power
in a numerical way with the same information shown in the previous two plots.

\newpage

# Run Predictions
With the same dataset, to run a prediction, a different parameter is needed:

- `rangeSimNumRep`: the power of replicate number 5 to 20 will be predicted.

Similar to the estimation process, however, the simulations will be excuted with
each sample size specified in `rangeSimNumRep`. (Note: the term sample size in
this vignette refers to the replicate number of each group/case)

It is possible to append the prediction results within the same object by using
the same result object as an input.
```{r prediction run2, message=FALSE, warning=FALSE}
data("exampleRNASeqData")
res <- predictPower(inputObject = res,
                    groupVec = exampleRNASeqData$groupVec,
                    isLogTransformed = FALSE,
                    dataType = "RNASeq",
                    rangeSimNumRep = c(5, 10, 15, 20),
                    minLFC = 1,
                    alpha = 0.05,
                    ST = 30,
                    seed = 345)
```

\newpage

## Visualization

The predicted results can be summaried using `plotPredPwr`. The input should be
the predicted power object returned from `predictPower`, the summary can be
optionally visualized by setting the following parameters:

- `plotType`: power-samplesize-foldchange relationship can be visualized
optionally between "lineplot" and "heatmap".
- `minLFC` and `maxLFC`: to observe power in a specific range of LFC
- `LFCscale`: to determine the LFC scale of the observation

### Line Plot

Lineplot (LFCscale = 0.5):
```{r LinePlot}
linePlotPredPwr(res, LFCscale = 0.5)
```

Lineplot is one of the optional outputs of `plotPredPwr`, the output contains a
lineplot and a summary table. For each comparison, the lineplot shows the power
tendency across every Log2 Fold Change segment resulted from a complete LFC list
divided by a specified `LFCscale`. Each dot on the lines stands for the average
power (y-axis) of the genes within the LFC range (x-axis), and each colour
indicates the average power of a certain sample size (as shown in the legend
besides the plot). In addition, a summary table below displays the average
power of each comparison across the sample sizes.

For instance, the line plot here shows the average power of four sample sizes
(5 to 30, with increment of 5) in LFCscale of 0.5. The LFC ranges from 0 to 5,
and within each LFC segment, the graph shows the average power of the features.
Here, the higher LFC shows higher power, the average power of each LFC range
increases with the larger sample sizes, as expected.

### Heatmap

Heatmap (LFCscale = 0.5):
```{r Heatmap}
heatMapPredPwr(res, LFCscale = 0.5)
```

The heatmap option presents the power predictions in a similar way, as previous
visualizations. Each heatmap displays overall power of each LFC range and sample
size. The average power of each LFC range is scaled with colours between blue
and red, as shown in the colour bar on the right. For example, this graph shows
the power increasing with larger sample sizes as expected. The same summary
table is also shown at the bottom.
