---
title: "PowerExplorer Manual"
author: "Xu Qiao"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{PowerExplorer Manual}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
## Abstract
This vignette demonstrates the applications of R package `PowerExplorer` as the power and sample size estimation tool for RNA-Seq and quantitative proteomics data. 

`PowerExplorer` contains following main features:

- Estimation of power based on the current setting
- Prediction of power according to the future settings (e.g. increasing sample size)
- Visualizations of estimation and prediction results

## Introduction
Power and sample size estimation is still one of the important principles in designing next-generation sequencing experiments to discover differential expressions, while power estimation method specialized for proteomics data has not yet developed.

Here we present `PowerExplorer`, which is a power estimation and prediction tool currently applicable to RNA-Seq and quantitative proteomics experiments. The calculation procedure starts with the estimation of distribution parameters of the each gene or protein (following referred as entry for simplicity) accordingly, with the obtained prior distribution of each entry, a specified amount of simulations are executed to repetitively generate data (read counts for RNA-Seq and peptide abundance for proteomics) for each entry based on null and alternative hypotheses respectively, furthermore, the corresponding statistical tests (t-test or Wald-test) are performed for each hypothesis, eventually from which result statistics will be summarized to calculate the statistical power.

\newpage

## Prepare Input Data

For datasets of both RNA-Seq (gene expression levels) and quantitative proteomics (peptide abundance levels), the data matrix should be arranged as entries in rows and samples in columns, for example:
```{r showDataExample, message=FALSE, warning=FALSE}
library(PowerExplorer)
data("exampleRNASeqData")
head(exampleRNASeqData$dataMatrix[,1:6])
```

A grouping vector indicating the sample groups to which all the samples belong should also be created, for example:
```{r}
show(exampleProteomicsData$groupVec)
```
```{r}
colnames(exampleProteomicsData$dataMatrix)
```

Note that the grouping vector length should be equal to the column number of the data matrix, all groups conventionally should have the same number of samples, otherwise the tool will automatically even all the sample numbers to the least number to achieve equal groups.

\newpage

## Run Estimation

Here we use a randomly generated proteomics dataset `exampleProteomicsData` as an example to estimate the current power of the dataset. The input dataset is named as `dataMatrix` and the grouping vector as `groupVec`.

To run the estimation, apart from the input, we still need to specify the following parameters:

- `isLogTransformed`: FALSE, the input data is not log-transformed
- `dataType`: "Proteomics", declare the datatype as "Proteomics"
- `minLFC`: 0.5, the threshold of Log2 Fold Change, proteins with lower LFC will be discarded
- `alpha`: 0.05, the controlled false positive (Type I Error) rate
- `ST`: 50, the simulation of each protein entry will be run 50 times (ST>50 is recommended)
- `seed`: 345, the seed of the random variables to maintain the reproducibility
- `showProcess`: FALSE, no detailed processes will be shown, set to TRUE if debug is needed
- `saveSimulatedData`: TRUE, save the simulated data in ~/savedData directory

The results will be summaried in barplot, boxplot and summary table.

```{r estimation run, fig.keep='none', message=FALSE, warning=FALSE, results='hide'}
library(PowerExplorer)
data("exampleProteomicsData")
estimatedPower <- estimateCurrentPower(inputDataMatrix = exampleProteomicsData$dataMatrix, 
                                       groupVec = exampleProteomicsData$groupVec, 
                                       isLogTransformed = FALSE, 
                                       dataType = "Proteomics", 
                                       minLFC = 0.5, 
                                       alpha = 0.05, 
                                       ST = 50, 
                                       seed = 345, 
                                       showProcess = FALSE, 
                                       saveSimulatedData = FALSE)
```

The estimated results can be summarized using `plotEstimatedPower`, the input data should be the estimated power object produced from  `estimateCurrentPower`.

```{r plot estimated power, echo = FALSE}
plotEstimatedPower(powerList = estimatedPower)
```

\newpage

## Run Predictions
With the same dataset, to run a prediction, a few more parameters are needed:

- `isLogTransformed`: FALSE, the input data is not log-transformed
- `dataType`: "Proteomics", declare the datatype as "Proteomics"
- `rangeSimNumRep`: the power of replicate number 5 to 20 will be predicted
- `alpha`: 0.05, the controlled false positive rate
- `ST`: 30, the statistical test and data simulation of each protein entry will be run 30 times(ST>50 is recommended)
- `seed`: 345, specify the seed of the random variables to maintain the reproducibility
- `showProcess`: FALSE, no detailed processes will be shown, set to TRUE if debug is needed
- `saveSimulatedData`: TRUE, save the simulated data in root/savedData directory

Similar to the estimation process, however, the simulations will be excuted with each sample size specified in `rangeSimNumRep`. (Note: the term sample size in this vignette refers to the replicate number of each group/case)

```{r prediction run, fig.keep='none', message=FALSE, warning=FALSE, results='hide'}
data("exampleProteomicsData")
predictedPower <- predictSampleSizePower(inputDataMatrix = exampleProteomicsData$dataMatrix,
                                         groupVec = exampleProteomicsData$groupVec,
                                         isLogTransformed = FALSE,
                                         dataType = "Proteomics",
                                         rangeSimNumRep = c(5, 10, 15, 20),
                                         alpha = 0.05,
                                         ST = 30,
                                         seed = 345,
                                         showProcess = FALSE,
                                         saveSimulatedData = FALSE)
```

The predicted results can be summaried using `plotPredictedPower`. The input should be the predicted power object returned from `predictSampleSizePower`, the summary can be optionally visualized by setting the following parameters:

- `plotType`: power-samplesize-foldchange relationship can be visualized optionally between "lineplot" and "heatmap".
- `minLFC` and `maxLFC`: to observe power in a specific range of LFC
- `LFCscale`: to determine the LFC scale of the observation

\newpage

Lineplot (LFCscale = 0.5):
```{r LinePlot, fig.height=3.5}
plotPredictedPower(predictedPower = predictedPower,
                   plotType = "lineplot",
                   LFCscale = 0.5)
```

Heatmap (LFCscale = 0.5):
```{r Heatmap, fig.height=3.5}
plotPredictedPower(predictedPower = predictedPower,
                   plotType = "heatmap",
                   LFCscale = 0.5)
```
