---
title: "PowerExplorer Manual"
author: "Xu Qiao"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
toc: true
vignette: >
  %\VignetteIndexEntry{PowerExplorer Manual}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

\newpage

# Abstract
This vignette demonstrates the applications of R package `PowerExplorer` as the power and sample size estimation tool for RNA-Seq and quantitative proteomics data. 

`PowerExplorer` contains following main features:

- Estimation of power based on the current setting
- Prediction of power according to the future settings (e.g. increasing sample size)
- Visualizations of estimation and prediction results

# Introduction
Power and sample size estimation is still one of the important principles in designing next-generation sequencing experiments to discover differential expressions, a few methods on power estimation for RNA-Seq data have been studied, while the one specialized for proteomics data has not yet been developed. `PowerExplorer`is a power estimation and prediction tool currently applicable to RNA-Seq and quantitative proteomics experiments. The calculation procedure starts with estimating the distribution parameters of each gene or protein (following referred as entry for simplicity) accordingly, with the obtained prior distribution of each entry, a specified amount of simulations are executed to generate data (read counts for RNA-Seq and peptide abundance for proteomics) repetitively for each entry based on null and alternative hypotheses. Furthermore, the corresponding statistical tests (t-test or Wald-test) are performed and the test statistics are collected, eventually the result statistics will be summarized to calculate the statistical power.

\newpage

# Prepare Input Data

For both RNA-Seq (gene expression levels) and quantitative proteomics (peptide abundance levels) datasets, the data matrix should be arranged as entries in rows and samples in columns, for example:
```{r showDataExample, message=FALSE, warning=FALSE}
library(PowerExplorer)
data("exampleRNASeqData")
head(exampleRNASeqData$dataMatrix[,1:6])
```

A grouping vector indicating the sample groups to which all the samples belong should also be created, for example:
```{r}
show(exampleProteomicsData$groupVec)
```
```{r}
colnames(exampleProteomicsData$dataMatrix)
```

Note that the grouping vector length should be equal to the column number of the data matrix, all groups conventionally should have the same number of samples, otherwise the tool will automatically even all the sample numbers to the least number to achieve equal groups.

\newpage

# Run Estimation

Here we use a randomly generated proteomics dataset `exampleProteomicsData` as an example to estimate the current power of the dataset. The input dataset is named as `dataMatrix` and the grouping vector as `groupVec`.

To run the estimation, apart from the input, we still need to specify the following parameters:

- `isLogTransformed`: FALSE; the input data is not log-transformed.
- `dataType`: "RNA-Seq"; the datatype can be declared as "Proteomics" or "RNA-Seq".
- `minLFC`: 0.5; the threshold of Log2 Fold Change, proteins with lower LFC will be discarded.
- `alpha`: 0.05; the controlled false positive (Type I Error) rate.
- `ST`: 50; the simulation of each protein entry will be run 50 times (ST>50 is recommended).
- `seed`: 345; the seed of the random variables to maintain the reproducibility.
- `showProcess`: FALSE; no detailed processes will be shown, set to TRUE if debug is needed.
- `saveSimulatedData`: FALSE; if TRUE, save the simulated data in ~/savedData directory.

The results will be summaried in barplot, boxplot and summary table.

```{r estimation run, fig.keep='none', message=FALSE, warning=FALSE, results='hide'}
library(PowerExplorer)
data("exampleRNASeqData")
estimatedPower <- estimateCurrentPower(inputDataMatrix = exampleRNASeqData$dataMatrix, 
                                       groupVec = exampleRNASeqData$groupVec, 
                                       isLogTransformed = FALSE, 
                                       dataType = "RNA-Seq", 
                                       minLFC = 0.5, 
                                       alpha = 0.05, 
                                       ST = 50, 
                                       seed = 345, 
                                       showProcess = FALSE, 
                                       saveSimulatedData = FALSE)
```

A part of the output should look like this:
```{r echo=FALSE, out.width = "90%"}
imgpath <- paste0(getwd(), "/vignettes/outputExample.png")
knitr::include_graphics(imgpath, auto_pdf = TRUE, dpi = NULL)
```

\newpage

## Visualization
The estimated results can be summarized using `plotEstimatedPower`, the only input needed is the `estimatedPower`, which should be the estimated power object returned from  `estimateCurrentPower`.

```{r plot estimated power, echo = FALSE}
plotEstimatedPower(powerList = estimatedPower)
```

The graph contains 3 plots, the `barplot` vertically shows the number of genes/proteins above the minLFC threshold, columns indicates the comparison pairs, each column presents the proportions of three power levels in three colours as indicated in the legend `power.level`; The boxplot shows the overall power distribution of each comparsion; And the summary table summarize the power in a numerical way with the same information shown in the previous two plots.

\newpage

# Run Predictions
With the same dataset, to run a prediction, a few more parameters are needed:

- `isLogTransformed`: FALSE; the input data is not log-transformed.
- `dataType`: "RNA-Seq"; the datatype can be declared as "Proteomics" or "RNA-Seq".
- `rangeSimNumRep`: the power of replicate number 5 to 20 will be predicted.
- `alpha`: 0.05; the controlled false positive rate.
- `ST`: 30; the statistical test and data simulation of each protein entry will be run 30 times (ST>50 is recommended).
- `seed`: 345; specify the seed of the random variables to maintain the reproducibility.
- `showProcess`: FALSE; no detailed processes will be shown, set to TRUE if debug is needed.
- `saveSimulatedData`: FALSE; if TRUE, save the simulated data in root/savedData directory.

Similar to the estimation process, however, the simulations will be excuted with each sample size specified in `rangeSimNumRep`. (Note: the term sample size in this vignette refers to the replicate number of each group/case)

```{r prediction run, eval=FALSE, include=FALSE, echo=TRUE}
data("exampleRNASeqData")
predictedPower <- predictSampleSizePower(inputDataMatrix = exampleRNASeqData$dataMatrix,
                                         groupVec = exampleRNASeqData$groupVec,
                                         isLogTransformed = FALSE,
                                         dataType = "RNA-Seq",
                                         rangeSimNumRep = c(5, 10, 15, 20),
                                         alpha = 0.05,
                                         ST = 30,
                                         seed = 345,
                                         showProcess = FALSE,
                                         saveSimulatedData = FALSE)
```

A part of the output should look like this:
```{r echo=FALSE, out.width = "90%"}
imgpath <- paste0(getwd(), "/vignettes/outputExample2.png")
knitr::include_graphics(imgpath, auto_pdf = TRUE, dpi = NULL)
```

\newpage

## Visualization

The predicted results can be summaried using `plotPredictedPower`. The input should be the predicted power object returned from `predictSampleSizePower`, the summary can be optionally visualized by setting the following parameters:

- `plotType`: power-samplesize-foldchange relationship can be visualized optionally between "lineplot" and "heatmap".
- `minLFC` and `maxLFC`: to observe power in a specific range of LFC
- `LFCscale`: to determine the LFC scale of the observation

### Line Plot

Lineplot (LFCscale = 0.5):
```{r LinePlot}
data("examplePredictedPower")
plotPredictedPower(examplePredictedPower, plotType = "lineplot", LFCscale = 0.5)
```

Lineplot is one of the optional outputs of `plotPredictedPower`, the output contains a lineplot and a summary table. For each comparison, the lineplot shows the power tendency across each Log2 Fold Change segment, which resulted from a complete LFC list divided by a specified `LFCscale`. Each dot on the lines stands for the average power (y-axis) of the genes within the LFC range (x-axis), and the colours indicate the average power of the certain sample size as shown in the legend besides the plot. In addition, a summary table below shows the average power of each comparison across the sample sizes. 

For instance, the line plot here shows the average power of four sample sizes (5 to 30, step=5) in LFCscale of 0.5, the LFC range is between 0 and 5, each LFC segment shows the average power of the entries with the LFC in this range, here the higher LFC has higher power, additionally as the repNum (sample size) shown with different colours, the average power of each LFC range increases with the larger sample sizes. 

### Heatmap

Heatmap (LFCscale = 0.5):
```{r Heatmap}
data("examplePredictedPower")
plotPredictedPower(examplePredictedPower, plotType = "heatmap", LFCscale = 0.5)
```

The heatmap option presents the power predictions in the similar way, vertically each heatmap shows overall LFC level of a comparison, sometimes a certain range shows blank space, since the result LFC vary in different comparison pairs. The average power of each LFC range is scaled with colours between blue and red, the middle value (0.6) is coloured as yellow, as shown in the colour bar on the right. For example, this graph shows the power increases with larger sample sizes. The same summary table is also shown on the bottom.

